presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -49,31 +83,186 @@ public static PartitionStatistics merge(PartitionStatistics first, PartitionStat
+    {
+        if (first.isPresent() && second.isPresent()) {
+            return Optional.of(new IntegerStatistics(
+                    reduce(first.get().getMin(), second.get().getMin(), SELECT_MIN, true),
+                    reduce(first.get().getMax(), second.get().getMax(), SELECT_MAX, true)));
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<DoubleStatistics> mergeDoubleStatistics(Optional<DoubleStatistics> first, Optional<DoubleStatistics> second)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            return Optional.of(new DoubleStatistics(
+                    reduce(first.get().getMin(), second.get().getMin(), SELECT_MIN, true),
+                    reduce(first.get().getMax(), second.get().getMax(), SELECT_MAX, true)));
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<DecimalStatistics> mergeDecimalStatistics(Optional<DecimalStatistics> first, Optional<DecimalStatistics> second)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            return Optional.of(new DecimalStatistics(
+                    reduce(first.get().getMin(), second.get().getMin(), SELECT_MIN, true),
+                    reduce(first.get().getMax(), second.get().getMax(), SELECT_MAX, true)));
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<DateStatistics> mergeDateStatistics(Optional<DateStatistics> first, Optional<DateStatistics> second)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            return Optional.of(new DateStatistics(
+                    reduce(first.get().getMin(), second.get().getMin(), SELECT_MIN, true),
+                    reduce(first.get().getMax(), second.get().getMax(), SELECT_MAX, true)));
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<BooleanStatistics> mergeBooleanStatistics(Optional<BooleanStatistics> first, Optional<BooleanStatistics> second)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            return Optional.of(new BooleanStatistics(
+                    reduce(first.get().getTrueCount(), second.get().getTrueCount(), ADD, false),
+                    reduce(first.get().getFalseCount(), second.get().getFalseCount(), ADD, false)));
+        }
+        return Optional.empty();
+    }
+
+    private static OptionalDouble mergeAverage(OptionalDouble first, OptionalLong firstRowCount, OptionalDouble second, OptionalLong secondRowCount)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            if (!(firstRowCount.isPresent() && secondRowCount.isPresent())) {
+                return OptionalDouble.empty();
+            }
+            long totalRowCount = firstRowCount.getAsLong() + secondRowCount.getAsLong();
+            if (totalRowCount == 0) {
+                return OptionalDouble.empty();
```

Comment:
If there 0 rows the average length is unknow. There is no rows. If we inserted values into empty partition it will go the `first.isPresent() ? first : second;` path.